qadence/noise/readout.py

```diff
@@ -0,0 +1,201 @@
+        if output_type == "torch"
+        else np.array(list(bitstring)).astype(int)
+    )
+
+
+def tensor_to_bitstring(bitstring: Tensor | np.array, output_type: str = "torch") -> str:
+    """
+    A helper function to convert torch.Tensor or numpy.array to bit strings.
+
+    Args:
+        bitstring: A torch.Tensor or numpy.array format of a bit string.
+
+    Returns:
+        A str out of the input bit string.
+    """
+    return (
+        "".join(list(map(str, bitstring.detach().tolist())))
+        if output_type == "torch"
+        else "".join(bitstring.astype("str"))
+    )
+
+
+def bit_flip(bit: Tensor, cond: Tensor) -> Tensor:
+    """
+    A helper function that reverses the states 0 and 1 in the bit string.
+
+    Args:
+        bit: A integer-value bit in a bitstring to be inverted.
+        cond: A Bool value of whether or not a bit should be flipped.
+
+    Returns:
+        The inverse value of the input bit
+    """
+    return torch.where(cond, torch.where(bit == 0, 1, 0), bit)
+
+
+def sample_to_matrix(sample: dict) -> Tensor:
+    """
+    A helper function that maps a sample dict to a bit string array.
+
+    Args:
+        sample: A dictionary with bit stings as keys and values
+        as their counts.
+
+    Returns: A torch.Tensor of bit strings n_shots x n_qubits.
+    """
+
+    return torch.concatenate(
+        list(
+            map(
+                lambda bitstring: torch.broadcast_to(
+                    bitstring_to_tensor(bitstring), [sample[bitstring], len(bitstring)]
+                ),
+                sample.keys(),
+            )
+        )
+    )
+
+
+def create_noise_matrix(
+    noise_distribution: torch.distributions, n_shots: int, n_qubits: int
+) -> Tensor:
+    """
+    A helper function that creates a noise matrix for bit string corruption.
+
+    NB: The noise matrix is not square, as all bits are considered independent.
+
+    Args:
+        noise_distribution: Torch statistical distribution one of Gaussian,
+        Uniform, of Poisson.
+        n_shots: Number of shots/samples.
+        n_qubits: Number of qubits
+
+    Returns:
+        A sample out of the requested distribution given the number of shots/samples.
+    """
+    # the noise_matrix should be available to the user if they want to do error correction
+    return noise_distribution.sample([n_shots, n_qubits])
+
+
+def bs_corruption(
+    err_idx: Tensor,
+    sample: Tensor,
+) -> Counter:
+    """
+    A function that incorporates the expected readout error in a sample of bit strings.
+
+    given a noise matrix.
+
+    Args:
+        err_idx: A Boolean array of bit string indices that need to be corrupted.
+        sample: A torch.Tensor of bit strings n_shots x n_qubits.
+
+    Returns:
+        A counter of bit strings after readout corruption.
+    """
+
+    func = torch.func.vmap(bit_flip)
+
+    return Counter([tensor_to_bitstring(k) for k in func(sample, err_idx)])
+
+
+def error(
```

Comment:
Name of the funciton is misleading, I think. I would call it `add_noise` for example (and also change the way it is called in the protocols enumeration then).